### ISSUE TITLE
Grafana causes deadlocks in PostgreSQL, while trying to refresh users token.

### ISSUE THREAD

**Issue by hgranlund** [grafana/grafana#20515]
**What happened**:
Grafana failed to refresh token due to a deadlock in PostgreSQL. The deadlock happend while grafana tried to do multiple updates to the "user_auth"-table.

***grafana.log***
```
msg="Failed to update access token during token refresh" logger=data-proxy-log error="pq: deadlock detected"
```
***postgresql.log***
```
2019-11-20 10:33:03 UTC-5dd5107d.8c20c-ERROR:  deadlock detected
2019-11-20 10:33:03 UTC-5dd5107d.8c20c-DETAIL:  Process 573964 waits for ShareLock on transaction 860860; blocked by process 571072.
	Process 571072 waits for ShareLock on transaction 860862; blocked by process 568396.
	Process 568396 waits for ExclusiveLock on tuple (81,3) of relation 21993 of database 21484; blocked by process 573964.
	Process 573964: UPDATE "user_auth" SET "user_id" = $1, "auth_module" = $2, "created" = $3, "o_auth_access_token" = $4, "o_auth_refresh_token" = $5, "o_auth_token_type" = $6, "o_auth_expiry" = $7 WHERE "user_id"=$8 AND "auth_module"=$9
	Process 571072: UPDATE "user_auth" SET "user_id" = $1, "auth_module" = $2, "created" = $3, "o_auth_access_token" = $4, "o_auth_refresh_token" = $5, "o_auth_token_type" = $6, "o_auth_expiry" = $7 WHERE "user_id"=$8 AND "auth_module"=$9
	Process 568396: UPDATE "user_auth" SET "user_id" = $1, "auth_module" = $2, "created" = $3, "o_auth_access_token" = $4, "o_auth_refresh_token" = $5, "o_auth_token_type" = $6, "o_auth_expiry" = $7 WHERE "user_id"=$8 AND "auth_module"=$9

```

**What you expected to happen**:

Grafana should not cause deadlocks in PostgreSQL

**How to reproduce it (as minimally and precisely as possible)**:

Set up Grafana with:

* A postgres instance. (We use a PostgreSQL DBaaS instance with network latency  around 5ms).
* A generic oauth provider
* A datasource with “Forward OAuth Identity” enabled.
* Set the [token_rotation_interval_minutes](https://grafana.com/docs/auth/overview/#settings) to as low as possible, as this bug happends while the token is being refreshed.
* Create a dashboard that do multiple requests to the datasource.

Set the dashboard to refresh every 5s, and wait for the deadlock to appear.

**Anything else we need to know**:

I have collected logs from grafana-server and PostgreSQL, as well as fetched the grafana.ini file:

[grafana.log](https://github.com/grafana/grafana/files/3869286/grafana.log)
[postgresql.log](https://github.com/grafana/grafana/files/3869287/postgresql.log)
[grafana.ini.txt](https://github.com/grafana/grafana/files/3869290/grafana.ini.txt)

**Environment**:
Grafana version: 6.3.4 (docker-image: grafana/grafana:6.3.4)
PostgreSQL version 11

**Comment by marefr** [comment]
> Set the token_rotation_interval_minutes to as low as possible, as this bug happends while the token is being refreshed.

I don't think that this should matter because the Grafana internal auth token rotation happens in the beginning of the request cycle.

Seem to me like this feature wasn't implemented/accepted with high load in mind. To fix this we would either implement #12131 or either move the rotation of OAuth access tokens to a background process.

**Comment by hgranlund** [comment]
> Set the token_rotation_interval_minutes to as low as possible, as this bug happends while the token is being refreshed.
> 
> I don't think that this should matter because the Grafana internal auth token rotation happens in the beginning of the request cycle.

You may be right. My theory is that with normal requests, without token refresh, all updates of the "user_auth" table will be the same. Which again, probably won't cause a deadlock in PostgreSQL? However, if the token is rotated, all updates will be different and a deadlock will be more likely. I am no PostgreSQL expert, so my theory may be wrong.


> Seem to me like this feature wasn't implemented/accepted with high load in mind. To fix this we would either implement #12131 or either move the rotation of OAuth access tokens to a background process.

Of the two solutions you mentioned, I most believe in the second one, to move the rotation of OAuth access tokens to a background process.
Is there an issue on it? And is this a feature that is wanted by the grafana community?

**Comment by marefr** [comment]
It's this issue and #12131 - should be enough. #12131 is something on our backlog that we plan to work on soon hopefully.

**Comment by papagian** [comment]
@hgranlund do you still use the specific database and experience similar deadlocks?
There was an [issue](https://github.com/grafana/grafana/issues/23077) with duplicated entries in the database that it could be related. I wonder whether you also had such duplicates.

**Comment by mildwonkey** [comment]
I think this issue is resolved and can be closed. Various fixes have already been merged since this was opened, and there haven't been any  new report or comments on this ticket in over a year. 

I tried - and failed - to reproduce this locally using postgres as the backend database, keycloak for oauth, a graphite datasource with oath forwarding, and a dashboard with a couple of panels on 5s refresh. I've seen no issues with token rotation.

I am going to close this issue, though we can re-open if there are objections. If someone is still experiencing this behavior in the latest version of Grafana, please open a new issue (since the underlying code base has changed quite a bit) with detailed reproduction steps. Thanks!
